batch/sql/improve-cancel.sql

```diff
@@ -0,0 +1,567 @@
+
+  IF expected_attempt_id != in_attempt_id THEN
+    COMMIT;
+    SELECT 2 as rc,
+      expected_attempt_id,
+      delta_cores_mcpu,
+      'input attempt id does not match expected attempt id' as message;
+  ELSEIF cur_job_state = 'Ready' OR cur_job_state = 'Running' THEN
+    UPDATE jobs
+    SET state = new_state, status = new_status, attempt_id = NULL
+    WHERE batch_id = in_batch_id AND job_id = in_job_id;
+
+    UPDATE batches SET n_completed = n_completed + 1 WHERE id = in_batch_id;
+    UPDATE batches
+      SET time_completed = new_timestamp,
+          `state` = 'complete'
+      WHERE id = in_batch_id AND n_completed = batches.n_jobs;
+
+    IF new_state = 'Cancelled' THEN
+      UPDATE batches SET n_cancelled = n_cancelled + 1 WHERE id = in_batch_id;
+    ELSEIF new_state = 'Error' OR new_state = 'Failed' THEN
+      UPDATE batches SET n_failed = n_failed + 1 WHERE id = in_batch_id;
+    ELSE
+      UPDATE batches SET n_succeeded = n_succeeded + 1 WHERE id = in_batch_id;
+    END IF;
+
+    UPDATE jobs
+      INNER JOIN `job_parents`
+        ON jobs.batch_id = `job_parents`.batch_id AND
+           jobs.job_id = `job_parents`.job_id
+      SET jobs.state = IF(jobs.n_pending_parents = 1, 'Ready', 'Pending'),
+          jobs.n_pending_parents = jobs.n_pending_parents - 1,
+          jobs.cancelled = IF(new_state = 'Success', jobs.cancelled, 1)
+      WHERE jobs.batch_id = in_batch_id AND
+            `job_parents`.batch_id = in_batch_id AND
+            `job_parents`.parent_id = in_job_id;
+
+    COMMIT;
+    SELECT 0 as rc,
+      cur_job_state as old_state,
+      delta_cores_mcpu;
+  ELSEIF cur_job_state = 'Cancelled' OR cur_job_state = 'Error' OR
+         cur_job_state = 'Failed' OR cur_job_state = 'Success' THEN
+    COMMIT;
+    SELECT 0 as rc,
+      cur_job_state as old_state,
+      delta_cores_mcpu;
+  ELSE
+    COMMIT;
+    SELECT 1 as rc,
+      cur_job_state,
+      delta_cores_mcpu,
+      'job state not Ready, Running or complete' as message;
+  END IF;
+END $$
+
+DROP TRIGGER IF EXISTS jobs_after_update;
+CREATE TRIGGER jobs_after_update AFTER UPDATE ON jobs
+FOR EACH ROW
+BEGIN
+  DECLARE cur_user VARCHAR(100);
+  DECLARE cur_batch_cancelled BOOLEAN;
+  DECLARE cur_n_tokens INT;
+  DECLARE rand_token INT;
+
+  SELECT user, cancelled INTO cur_user, cur_batch_cancelled FROM batches
+  WHERE id = NEW.batch_id;
+
+  SELECT n_tokens INTO cur_n_tokens FROM globals LOCK IN SHARE MODE;
+  SET rand_token = FLOOR(RAND() * cur_n_tokens);
+
+  IF OLD.state = 'Ready' THEN
+    IF NOT (OLD.always_run OR OLD.cancelled OR cur_batch_cancelled) THEN
+      # cancellable
+      INSERT INTO batch_cancellable_resources (batch_id, token, n_ready_cancellable_jobs, ready_cancellable_cores_mcpu)
+      VALUES (OLD.batch_id, rand_token, -1, -OLD.cores_mcpu)
+      ON DUPLICATE KEY UPDATE
+	n_ready_cancellable_jobs = n_ready_cancellable_jobs - 1,
+	ready_cancellable_cores_mcpu = ready_cancellable_cores_mcpu - OLD.cores_mcpu;
+    END IF;
```

Comment:
Don't you need to update the ready cores in here as well?

Reply:
I'm not sure where "here" is.  Can you be more specific?

ready_cores is really runnable cores (I'm debating renaming).  Remember, runnable cores are ready cores that are not cancelled:

 > jobs.state = 'Ready' AND (jobs.always_run OR NOT (jobs.cancelled OR batches.cancelled))

so ready_cores are updated in the `state = 'Ready'` section (the first section) when the job is not cancelled:

```
    ELSE
      # runnable
      INSERT INTO user_resources (user, token, n_ready_jobs, ready_cores_mcpu)
      VALUES (cur_user, rand_token, -1, -OLD.cores_mcpu)
      ON DUPLICATE KEY UPDATE
        n_ready_jobs = n_ready_jobs - 1,
        ready_cores_mcpu = ready_cores_mcpu - OLD.cores_mcpu;

      INSERT INTO ready_cores (token, ready_cores_mcpu)
      VALUES (rand_token, -OLD.cores_mcpu)
      ON DUPLICATE KEY UPDATE
        ready_cores_mcpu = ready_cores_mcpu - OLD.cores_mcpu;
    END IF;
```

Reply:
"here" is the case for "cancellable".

When I looked at the trigger, this is your definition of cancellable: `NOT (OLD.always_run OR OLD.cancelled OR cur_batch_cancelled)`. I expanded that out to `NOT OLD.always_run AND NOT OLD.cancelled AND NOT cur_batch_cancelled`. To me that's not cancelled and should be counted as ready_cores. In this case, these "cancellable" cores should be subtracted from the ready cores.

I'll look at it again later.

Reply:
This is in the `state = 'Running'` section.  Running jobs should not be counted as ready.

Reply:
Ok. Going from my interpretation of the definitions above, if runnable and cancelled are opposites, then the ready cores update makes sense.